### PR TITLE
Anomaly Detection

### PR DESCRIPTION
**Pull request by KilnOfTheSecondFlame**

### CODE REVIEW

File: FS19/ML/Zusammenfassung_ML.tex [hunk 1]
```diff
@@ -1888,5 +1888,59 @@ \subsubsection{Lift of an Association Rule}
 	Lift = Association Strength
 \end{theorem}
 
+\section{Anomaly or Outlier Detection}
+
+Anomaly or Outlier Detection is an important topic in unsupervised learning. Anomaly and outlier are used as synonyms. Anomalies are patterns in data that do not conform to a well-defined notion of normal behavior.
+
+\begin{figure}[tbh!]
+    \centering
+    \includegraphics[width=0.4\linewidth, keepaspectratio]{Pictures/outlier_examples}
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:1897]: FYI, 'Pictures' is already in the picture path, no need to put that there

File: FS19/ML/Zusammenfassung_ML.tex [hunk 2]
```diff
@@ -1888,5 +1888,59 @@ \subsubsection{Lift of an Association Rule}
...
+    \item \textbf{Contextual outliers} deviate significantly with respect to a specific context
+    \item \textbf{Collective outliers} deviate as a group from the entire dataset but not necessarily as individuals
+\end{enumerate}
+
+We can solve the problem of outlier detection with machine learning:
+
+\begin{itemize}[leftmargin=*, labelindent=5cm, labelsep=0.5cm]
+    \item[\textbf{Supervised Learning}] Outlier detection can be considered a classification problem with categories \emph{normal} and \emph{outlier}
+        When outliers are rare events, collecting enough training data becomes a severe issue, and special measures must be taken because the two classes are extremely disbalanced
+    \item[\textbf{Unsupervised learning}] These methods assume that outliers deviate from the structural pattern of normal data objects.
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: Unsupervised Learning

File: FS19/ML/Zusammenfassung_ML.tex [hunk 3]
```diff
@@ -1888,5 +1888,59 @@ \subsubsection{Lift of an Association Rule}
...
+    \item \textbf{Global outliers} deviate significantly from the rest of the dataset
+    \item \textbf{Contextual outliers} deviate significantly with respect to a specific context
+    \item \textbf{Collective outliers} deviate as a group from the entire dataset but not necessarily as individuals
+\end{enumerate}
+
+We can solve the problem of outlier detection with machine learning:
+
+\begin{itemize}[leftmargin=*, labelindent=5cm, labelsep=0.5cm]
+    \item[\textbf{Supervised Learning}] Outlier detection can be considered a classification problem with categories \emph{normal} and \emph{outlier}
+        When outliers are rare events, collecting enough training data becomes a severe issue, and special measures must be taken because the two classes are extremely disbalanced
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:1914]: Add a linebreak between `\emph{outlier}` and _When_

File: FS19/ML/Zusammenfassung_ML.tex [hunk 4]
```diff
@@ -1888,5 +1888,59 @@ \subsubsection{Lift of an Association Rule}
...
+    \item[\textbf{Unsupervised learning}] These methods assume that outliers deviate from the structural pattern of normal data objects.
+    \begin{enumerate}
+        \item Statistical Methods
+        \item Proximity-based Methods
+        \item Clustering Methods
+    \end{enumerate}
+\end{itemize}
+
+\subsection{Statistical Methods}
+Statistical methods assume that normal data objects are generated by a stochastic model, and that data not following this model are outliers.
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: [...] and that data not following _that_ model are outliers.

File: FS19/ML/Zusammenfassung_ML.tex [hunk 5]
```diff
@@ -1888,5 +1888,59 @@ \subsubsection{Lift of an Association Rule}
...
+
+\begin{enumerate}
+    \item Assumption: The data follows a normal distribution with parameters $\mu$ and $\sigma^2$
+    \item Method: Find best parameters with Maximum Likelihood method
+    \item Conclusion: Find threshold were most of the data is located, for example $\mu \pm 3\sigma$ and classify all data outside as outliers
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: Find threshold where [...] classify all data _outside of this range_ as outliers.

File: FS19/ML/Zusammenfassung_ML.tex [hunk 6]
```diff
@@ -1888,5 +1888,59 @@ \subsubsection{Lift of an Association Rule}
...
+
+\begin{figure}[tbh!]
+    \centering
+    \includegraphics[width=0.5\linewidth, keepaspectratio]{Pictures/outlier_boxplot}
+    \caption{Comparison of BoxPlots to Normal Distribution}
+    \label{fig:outlierboxplot}
+\end{figure}
+
+Observe that this techniques for anomaly detection consider only single variables.
+% TODO add multiple variable anomaly detection from exercises
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:1940]: Fix ToDo

File: FS19/ML/Zusammenfassung_ML.tex [hunk 7]
```diff
@@ -1888,5 +1888,59 @@ \subsubsection{Lift of an Association Rule}
...
+
+\begin{enumerate}
+    \item Assumption: The data follows a normal distribution with parameters $\mu$ and $\sigma^2$
+    \item Method: Find best parameters with Maximum Likelihood method
+    \item Conclusion: Find threshold were most of the data is located, for example $\mu \pm 3\sigma$ and classify all data outside as outliers
+\end{enumerate}
+
+\begin{figure}[tbh!]
+    \centering
+    \includegraphics[width=0.5\linewidth, keepaspectratio]{Pictures/outlier_boxplot}
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: Bigger Picture, font is unreadable without magnifying

File: FS19/ML/Zusammenfassung_ML.tex [hunk 8]
```diff
@@ -1941,6 +1941,117 @@ \subsection{Statistical Methods}
 
 \subsection{Proximity-based Methods}
 
+In proximity-based methods records that are far from others are considered outliers. There are two types of proximity-based methods:
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: Shouldn't there be a comma? In proximity-based methods **,** records that are....

File: FS19/ML/Zusammenfassung_ML.tex [hunk 9]
```diff
@@ -1941,6 +1941,117 @@ \subsection{Statistical Methods}
...
+\begin{enumerate}
+    \item \textbf{Distance based methods} find outliers with respect to global dataset
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: _with respect to **the** global dataset_ or _with respect to global dataset**s**_

File: FS19/ML/Zusammenfassung_ML.tex [hunk 10]
```diff
@@ -1941,6 +1941,117 @@ \subsection{Statistical Methods}
...
+    \item \textbf{Density-based methods} find outliers with respect to local neighbourhood
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:1947]: Same as with dataset. add a _the_ or pluralify(?) neighbourhood

File: FS19/ML/Zusammenfassung_ML.tex [hunk 11]
```diff
@@ -1941,6 +1941,117 @@ \subsection{Statistical Methods}
...
+    \centering
+    \includegraphics[width=0.6\linewidth, keepaspectratio]{Pictures/outlier_proximity-based}
+    \caption{Outliers classified by proximity-based methods}
+    \label{fig:outlierproximity-based}
+\end{figure}
+
+\subsubsection{Distance-Based Ouliers}
+
+\begin{itemize}
+    \item Specify a ditsance threshold $r>0$ and a fraction threshold $0<\pi\leq 1$
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: Specify a _distance_ [...]

File: FS19/ML/Zusammenfassung_ML.tex [hunk 12]
```diff
@@ -1941,6 +1941,117 @@ \subsection{Statistical Methods}
...
+    \item For each record count the numbers of other records in the $r$-neighbourhood
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:1961]: comma between _record_ and _count_?

File: FS19/ML/Zusammenfassung_ML.tex [hunk 13]
```diff
@@ -1941,6 +1941,117 @@ \subsection{Statistical Methods}
...
+    \includegraphics[width=0.6\linewidth, keepaspectratio]{Pictures/outlier_density-based}
+    \caption{Graphical representation of density-based outlier detection}
+    \label{fig:outlierdensity-based}
+\end{figure}
+
+The local outlier factor (LOF) is a score that indicates how likely it is that a certain data point is an outlier. A $\text{LOF} \approx 1$ means no outlier, while $\text{LOF} \gg 1$ means outlier. The k-Distance is the distance of a point to its k-th neighbour.
+
+\paragraph{Reachability Distance}
+
+The reachability distance is simply the maximum of the distance of two points and the k-distance of the second point, consider it as a smoothing term, a lower distance bound that depends on the neighbourhood density.
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: The reachability distance is ~~simply~~ the maximum of the distance _between_ two points. Consider [...]

File: FS19/ML/Zusammenfassung_ML.tex [hunk 14]
```diff
@@ -1941,6 +1941,117 @@ \subsection{Statistical Methods}
...
+\begin{figure}[tbh!]
+    \centering
+    \includegraphics[width=0.5\linewidth, keepaspectratio]{Pictures/outlier_reachability_distance}
+    \caption{Reachability distance visualised}
+    \label{fig:outlierreachabilitydistance}
+\end{figure}
+
+\paragraph{Local Reachability Density (lrd)}
+
+Calculate the reachability distance of a to all its k-nearest neighbours and take the average of that number, get the density by taking the inverse.  Intuitively, local reachability density tells how far we have to travel to reach the next point or cluster of points, the lower the local reachability density of a point, the less dense its region, the longer we have to travel.
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: [...] of $a$ to all its [...] that number, _then_ get the density [...]
Intuitively, _the_ local reachability density tells _you_ how far we have to travel to reach the next point or cluster of points. The lower [...]

File: FS19/ML/Zusammenfassung_ML.tex [hunk 15]
```diff
@@ -1941,6 +1941,117 @@ \subsection{Statistical Methods}
...
+\begin{figure}[tbh!]
+    \centering
+    \includegraphics[width=0.5\linewidth, keepaspectratio]{Pictures/outlier_reachability_density}
+    \caption{The local reachability density is not symmetric}
+    \label{fig:outlierreachabilitydensity}
+\end{figure}
+
+\paragraph{Local Outlier Factor}
+
+Local reachability density of a point in comparison to its k neighbours, more specifically, $k$ ratios of the local reachability distance of each point to its neighbouring points are calculated and averaged.
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex]: _The Local Outlier Factor is the_ local reachability density of a point in comparison to its k neighbours. More specifically, _it's_ $k$ ratios of the local [...]

Also what does 'it's' refer to in this case?

File: FS19/ML/Zusammenfassung_ML.tex [hunk 16]
```diff
@@ -1941,6 +1941,117 @@ \subsection{Statistical Methods}
...
+\end{equation}
+
+\begin{figure}[tbh!]
+    \centering
+    \includegraphics[width=0.5\linewidth, keepaspectratio]{Pictures/outlier_reachability_density}
+    \caption{The local reachability density is not symmetric}
+    \label{fig:outlierreachabilitydensity}
+\end{figure}
+
+\paragraph{Local Outlier Factor}
```
Comment (gathurian) [inline FS19/ML/Zusammenfassung_ML.tex:2015]: ` \paragraph{Local Outlier Factor} \mbox{}\\`
To force a newline

**Review by gathurian — CHANGES_REQUESTED** [review]

File: FS19/ML/Zusammenfassung_ML.tex (hunk 6 above)
Comment (KilnOfTheSecondFlame) [reply FS19/ML/Zusammenfassung_ML.tex:1940]: Not yet
Will do when looking through exercises

File: FS19/ML/Zusammenfassung_ML.tex (hunk 1 above)
Comment (KilnOfTheSecondFlame) [reply FS19/ML/Zusammenfassung_ML.tex:1897]: TeXStudio does that automatically - IDC, as it does not break stuff =)

File: FS19/ML/Zusammenfassung_ML.tex (hunk 10 above)
Comment (KilnOfTheSecondFlame) [reply FS19/ML/Zusammenfassung_ML.tex:1947]: No as it's in the local neighbourhood, and respect to + Noun is valid
grafik

File: FS19/ML/Zusammenfassung_ML.tex (hunk 15 above)
Comment (KilnOfTheSecondFlame) [reply FS19/ML/Zusammenfassung_ML.tex]: No, it's literally k ratios (k in 1,2,3,4,5...)

File: FS19/ML/Zusammenfassung_ML.tex (hunk 16 above)
Comment (KilnOfTheSecondFlame) [reply FS19/ML/Zusammenfassung_ML.tex:2015]: I don't want a newline though =/

Comment (KilnOfTheSecondFlame) [reply FS19/ML/Zusammenfassung_ML.tex:2015]: grafik
Nice and compact in my opinion

**Review by gathurian — APPROVED** [review]

**Pull request merged**